R/coalesce.R

```diff
@@ -0,0 +1,34 @@
+coalesce <- function(x, ...) {
+  if (!anyNA(x) || missing(..1)) return(x)
```

Comment:
`anyNA` is not a good start, best to push that already to C where we have to test against NAs anyway.

Reply:
`anyNA` is Primitive, so it's very fast already. We should benchmark this & #3609 first

Reply:
It is fast because there is just a single simple loop which terminates as early as any NA is detected. In pessimistic scenario it goes through all rows, which is still relatively fast but unnecessary.

Reply:
unnecessary as in, best bet is probably not to bother with anyNA or count_na and just consistently pass over the vector once?

Reply:
added benchmarks, before proceeding with this optimisation it will be actually better to have a working version that we can refer to. I think we can get close to the floor of timing if we just do it in single pass. `anyNA(x)` is one pass, `sum(is.na(x))` is second and third, then our C loop is fourth.

Reply:
yep. the first 3 are primitive so they'll out perform our C loop individually but maybe not altogether

Reply:
890cd1a address this, see https://github.com/Rdatatable/data.table/pull/3608#issuecomment-496844042